pyro/infer/abstract_infer.py

```diff
@@ -285,23 +286,57 @@ class TracePredictive(TracePosterior):
     :param TracePosterior posterior: trace posterior instance holding
         samples from the model's approximate posterior.
     :param int num_samples: number of samples to generate.
+    :param list(str) resample_sites: sites to resample from in posterior predictive distribution.
+                                     By default all sites are used.
     """
-    def __init__(self, model, posterior, num_samples):
+    def __init__(self, model, posterior, num_samples, resample_sites=None):
         self.model = model
         self.posterior = posterior
         self.num_samples = num_samples
+        self.resample_sites = resample_sites
         super(TracePredictive, self).__init__()
 
     def _traces(self, *args, **kwargs):
         if not self.posterior.exec_traces:
             self.posterior.run(*args, **kwargs)
+        data_trace = poutine.trace(self.model).get_trace(*args, **kwargs)
         for _ in range(self.num_samples):
-            model_trace = self.posterior()
-            replayed_trace = poutine.trace(poutine.replay(self.model, model_trace)).get_trace(*args, **kwargs)
-            yield (replayed_trace, 0., 0)
+            model_trace = self.posterior().copy()
+            self._keep_sites(model_trace)
+            self._adjust_to_data(model_trace, data_trace)
+            resampler = poutine.resample_posterior(self.model, model_trace)
+            resampled_trace = poutine.trace(resampler).get_trace(*args, **kwargs)
+            yield (resampled_trace, 0., 0)
+
+    def _keep_sites(self, trace):
+        if self.resample_sites is None:
+            return
+        for name, site in list(trace.nodes.items()):
+            if name not in self.resample_sites:
+                trace.remove_node(name)
+
+    def _adjust_to_data(self, trace, data_trace):
+        for name, site in list(trace.nodes.items()):
+            # Adjust subsample sites
+            if site_is_subsample(site):
+                site["fn"] = data_trace.nodes[name]["fn"]
+                site["value"] = data_trace.nodes[name]["value"]
+            # Adjust sites under conditionally independent stacks
+            try:
+                site["cond_indep_stack"] = data_trace.nodes[name]["cond_indep_stack"]
+                site["fn"] = data_trace.nodes[name]["fn"]
+                for cis in site["cond_indep_stack"]:
+                    # Select random sub-indices to replay values under conditionally independent stacks.
+                    # Otherwise, we assume there is an dependence of indexes between training data
+                    # and prediction data.
+                    cat = Categorical(logits=site["value"].new_ones(site["value"].size(cis.dim)))
+                    subidxs = cat.sample((cis.size,))
```

Comment:
You can probably simplify this to `subidxs = Categorical(logits=site["value"].new_ones(cis.size)).sample()`.

Reply:
Thanks, will do!

Reply:
Actually, it seems that the change breaks the tests. 
The reason actually is that what I want is that the `logits` to only have one value, which is the number of possible elements from the training posterior and then the sampling from the categorical ensures that we select enough random elements for the prediction.

For example, let us say we trained with `500` samples, then I want it to be a `Categorical(500)`. The sample statement will the adjust itself to the prediction size, e.g. if we have `300` samples in our test data, we will pick 300 random indices between 0 and 499.

Reply:
Sorry, I had mistyped. I meant - `subidxs = Categorical(logits=site["value"].new_ones(cis.size)).sample([cis.size])`. Thanks for explaining though, that makes sense.

Reply:
That seems to work, thanks 😄 .

Reply:
Actually, the reason I am using the original is because I overwrite the context independent stack, so it will have the size of the evaluation data. But I want the categorical distribution to choose randomly according to the size of the training data (which I get from `site["value"].size(cis.dim)` instead). I will add a test where the evaluation data is larger than the training data as well 😄 .

Reply:
Thanks for explaining, this is getting really confusing! :)